bazel/external/quiche.BUILD

```diff
@@ -49,7 +49,11 @@ genrule(
     visibility = ["//visibility:private"],
 )
 
-quiche_copt = ["-Wno-unused-parameter"]
+quiche_copt = [
+    "-Wno-unused-parameter",
+    # QUIC uses offsetof() to optimize memory usage in frames.
+    "-Wno-invalid-offsetof",
```

Comment:
FWIW my patch of your PR compiles & tests fully with this line commented out.

Reply:
OK it's worth a comment. In general the awkward dependencies make it very hard for a reader of code in the Envoy tree to see why it's there, so commenting is really critical.

Better yet, can you change his hack in quiche to be inside a .cc file instead of a .h file so that this compile peculiarity is isolated, rather than viral.

Reply:
> OK it's worth a comment. In general the awkward dependencies make it very hard for a reader of code in the Envoy tree to see why it's there, so commenting is really critical.

There is a comment one line above: `QUIC uses offsetof() to optimize memory usage in frames.`

> Better yet, can you change his hack in quiche to be inside a .cc file instead of a .h file so that this compile peculiarity is isolated, rather than viral.

Saddly quic_inlined_frame.h uses class template. And definition of template class member has to be in the header file. A possible solution could be to move the method definition into .cc file together with declaration of `template class QuicInlinedFrame<FrameClass>` for all QuicFrame types, but this approach seems ugly too.

Reply:
Can you add some `#pragma` into the header, or a wrapper header, like I did for gRPC TSI headers: https://github.com/envoyproxy/envoy/blob/master/source/extensions/transport_sockets/alts/grpc_tsi.h#L7

It is more robust that the headers have the settings in themselves.

Reply:
The violating class in not supposed to be used by Envoy. And it's at the root of QUIC's dependency. So it's included by many quic core files. Currently only test files includes quic core files, but in the future every integration files will indirectly include the violating file. Are you suggesting add #pragma everywhere a quic core file is inclueded?

Even so there is still a difficulty where imported quiche test files, i.e. those included in test target quic_platform_api_test, can't be modified.  They are quiche test used to verify platform implementation correctness. There is not place to add pragma for them.

Reply:
@lizan that sounds like a good idea. @danzh2010 I did not understand your explanation of why the pragma could not be put into the file that needs it. This sounds like the preferred solution.

If it has to be here, can you add to the comment the fact that this only shows up with `bazel test @com_googlesource_quiche//...` and only in `quic_inlined_frame.h`, neither of which are in the Envoy tree. It's very difficult to understand why this potentially viral flag is needed in Envoy when nothing in the Envoy tree fails to compile without it, so explicit information is helpful.

Reply:
> @lizan that sounds like a good idea. @danzh2010 I did not understand your explanation of why the pragma could not be put into the file that needs it. 

We can put #pragma in files that depends on quic core files in the future. But for now, as said, @com_googlesource_quiche//... has test files brought in from quiche:
"quiche/quic/platform/api/quic_mem_slice_span_test.cc",
"quiche/quic/platform/api/quic_mem_slice_storage_test.cc",
"quiche/quic/platform/api/quic_mem_slice_test.cc"
These files can't be modified in envoy. So -Wno-invalid-offsetof is still needed for them.


> If it has to be here, can you add to the comment the fact that this only shows up with `bazel test @com_googlesource_quiche//...` and only in `quic_inlined_frame.h`, neither of which are in the Envoy tree. It's very difficult to understand why this potentially viral flag is needed in Envoy when nothing in the Envoy tree fails to compile without it, so explicit information is helpful.

For now not show in envoy tree, but in the future this error will show up in integration code, which can be suppressed using #pragma GCC diagnostic ignored as @lizan .

Reply:
I'm not following you. Are you going to do one of these options:
 * remove the -fno-invalid-offsetof
 * comment explicitly what breaks if you remove it

or something else?